File: test/ModularPipelines.UnitTests/Console/OutputCoordinatorDeferredFlushTests.cs
using Microsoft.Extensions.Logging;
using ModularPipelines.Context;
using ModularPipelines.Extensions;
using ModularPipelines.Modules;
using ModularPipelines.TestHelpers;

namespace ModularPipelines.UnitTests.Console;

public class OutputCoordinatorDeferredFlushTests : TestBase
{
    private class Module1 : Module<bool>
    {
        protected internal override async Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            context.Logger.LogInformation("Module1 output");
            await Task.Delay(50, cancellationToken);
            return true;
        }
    }

    private class Module2 : Module<bool>
    {
        protected internal override async Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            context.Logger.LogInformation("Module2 output");
            await Task.Yield();
            return true;
        }
    }

    [Test]
    public async Task Pipeline_Completes_When_Progress_Disabled()
    {
        var host = await TestPipelineBuilder.Create()
            .ConfigurePipelineOptions(options => options with
            {
                Console = options.Console with { ShowProgress = false },
            })
            .AddModule<Module1>()
            .BuildAsync();

        await using (host)
        {
            var result = await host.RunAsync();
            await Assert.That(result.Status).IsEqualTo(ModularPipelines.ModuleStatus.Succeeded);
        }
    }

    [Test]
    public async Task Pipeline_With_Multiple_Modules_Completes_Successfully()
    {
        var host = await TestPipelineBuilder.Create()
            .ConfigurePipelineOptions(options => options with
            {
                Console = options.Console with { ShowProgress = false },
            })
            .AddModule<Module1>()
            .AddModule<Module2>()
            .BuildAsync();

        await using (host)
        {
            var result = await host.RunAsync();
            await Assert.That(result.Status).IsEqualTo(ModularPipelines.ModuleStatus.Succeeded);
        }
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Console/OutputCoordinatorDeferredFlushTests.fs
namespace ModularPipelines.UnitTests.FSharp.Console

open System
open System.Threading.Tasks
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.Logging
open ModularPipelines.Context
open ModularPipelines.Enums
open ModularPipelines.Extensions
open ModularPipelines.Modules
open ModularPipelines.Options
open ModularPipelines.TestHelpers
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

type private Module1() =
    inherit Module<bool>()

    override _.ExecuteAsync(context: IModuleContext, cancellationToken) =
        task {
            context.Logger.LogInformation("Module1 output")
            do! Task.Delay(50, cancellationToken)
            return true
        }

type private Module2() =
    inherit Module<bool>()

    override _.ExecuteAsync(context: IModuleContext, _) =
        task {
            context.Logger.LogInformation("Module2 output")
            do! Task.Yield()
            return true
        }

type OutputCoordinatorDeferredFlushTests() =
    [<Test>]
    member _.Pipeline_Completes_When_Progress_Disabled() = async {
        let! host =
            TestPipelineHostBuilder.Create()
                .ConfigureServices(Action<ModularPipelines.PipelineBuilder, IServiceCollection>(fun _ services ->
                    services.Configure<PipelineOptions>(fun (options: PipelineOptions) ->
                        options.ShowProgressInConsole <- false)
                    |> ignore))
                .AddModule<Module1>()
                .BuildHostAsync()
            |> Async.AwaitTask

        let! result = host.ExecutePipelineAsync() |> Async.AwaitTask
        do! check(Assert.That(result.Status).IsEqualTo(Status.Successful))
        do! host.DisposeAsync().AsTask() |> Async.AwaitTask
    }

    [<Test>]
    member _.Pipeline_With_Multiple_Modules_Completes_Successfully() = async {
        let! host =
            TestPipelineHostBuilder.Create()
                .ConfigureServices(Action<ModularPipelines.PipelineBuilder, IServiceCollection>(fun _ services ->
                    services.Configure<PipelineOptions>(fun (options: PipelineOptions) ->
                        options.ShowProgressInConsole <- false)
                    |> ignore))
                .AddModule<Module1>()
                .AddModule<Module2>()
                .BuildHostAsync()
            |> Async.AwaitTask

        let! result = host.ExecutePipelineAsync() |> Async.AwaitTask
        do! check(Assert.That(result.Status).IsEqualTo(Status.Successful))
        do! host.DisposeAsync().AsTask() |> Async.AwaitTask
    }
